src/Storages/MergeTree/MergeTreeCurrentlyRestoringFromBackup.cpp

```diff
@@ -0,0 +1,333 @@
+
+        /// This `scope_guard` is used to remove the entry when we're done with the current RESTORE process.
+        scope_guard remove_entry = [this, entry_rawptr]
+        {
+            LOG_INFO(log, "Removing info about currently restoring parts");
+            std::lock_guard lock{mutex};
+            auto it = entries.find(entry_rawptr);
+            if (it != entries.end())
+                entries.erase(it);
+        };
+
+        {
+            /// Store the entry.
+            LOG_INFO(log, "Storing info about currently restoring parts");
+            std::lock_guard lock{mutex};
+            entries.emplace(std::move(entry));
+        }
+
+        return remove_entry;
+    }
+
+private:
+    StorageMergeTree & storage;
+    Poco::Logger * const log;
+
+    /// Represents parts and mutations restored by one RESTORE command to this table.
+    struct Entry
+    {
+        std::set<MergeTreePartInfo> part_infos;
+        std::unordered_set<Int64> mutation_numbers;
+    };
+
+    struct EntryPtrCompare
+    {
+        using is_transparent = void;
+        template <class P, class Q>
+        bool operator()(const P & left, const Q & right) const
+        {
+            return std::to_address(left) < std::to_address(right);
+        }
+    };
+
+    std::set<std::unique_ptr<Entry>, EntryPtrCompare> entries TSA_GUARDED_BY(mutex);
+    mutable std::mutex mutex;
+};
+
+
+std::optional<String> MergeTreeCurrentlyRestoringFromBackup::BlockNumbersAllocator::findAnyRestoringPart() const
+{
+    return currently_restoring_info.findAnyRestoringPart();
+}
+
+
+MergeTreeCurrentlyRestoringFromBackup::MergeTreeCurrentlyRestoringFromBackup(StorageMergeTree & storage_)
+    : log(&Poco::Logger::get(storage_.getLogName() + " (RestorerFromBackup)"))
+    , currently_restoring_info(std::make_unique<CurrentlyRestoringInfo>(storage_, log))
+    , block_numbers_allocator(std::make_unique<BlockNumbersAllocator>(storage_, log, *currently_restoring_info))
+{
+}
+
+MergeTreeCurrentlyRestoringFromBackup::~MergeTreeCurrentlyRestoringFromBackup() = default;
+
+bool MergeTreeCurrentlyRestoringFromBackup::containsPart(const MergeTreePartInfo & part_info) const
+{
+    return currently_restoring_info->containsPart(part_info);
+}
+
+bool MergeTreeCurrentlyRestoringFromBackup::containsMutation(Int64 mutation_number) const
+{
+    return currently_restoring_info->containsMutation(mutation_number);
+}
+
+scope_guard MergeTreeCurrentlyRestoringFromBackup::allocateBlockNumbers(
+    std::vector<MergeTreePartInfo> & part_infos_,
+    std::vector<MutationInfoFromBackup> & mutation_infos_,
+    bool check_table_is_empty_)
+{
+    auto check_for_no_parts_reason = getCheckForNoPartsReason(check_table_is_empty_, !mutation_infos_.empty());
+
+    /// Create temporary zookeeper nodes to allocate block numbers and mutation numbers.
+    block_numbers_allocator->allocateBlockNumbers(part_infos_, mutation_infos_, check_for_no_parts_reason);
```

Comment:
It's `MergeTree`, there are no ZooKeeper nodes

Reply:
I've fixed that copy-paste.